dnscache.go

```diff
@@ -96,12 +132,30 @@ func (r *Resolver) lookup(ctx context.Context, key string) (rrs []string, err er
 		if r.OnCacheMiss != nil {
 			r.OnCacheMiss()
 		}
-		rrs, err = r.update(ctx, key, true)
+		if r.concurrentLookups {
+			rrs, err = r.concurrentUpdate(ctx, key, true)
+		} else {
+			rrs, err = r.update(ctx, key, true)
+		}
 	}
 	return
 }
 
 func (r *Resolver) update(ctx context.Context, key string, used bool) (rrs []string, err error) {
+	rrs, err = r.doLookup(ctx, key)
+	if !r.saveErrors && err != nil {
+		return
+	}
+	if !r.saveEmpty && len(rrs) == 0 {
+		return
+	}
```

Comment:
Same comment here, please do only introduce M&S. Disabling negative caching should be discussed separately.